src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs

```diff
@@ -361,42 +391,56 @@ private void OnSourceProviderSourcesChanged(object sender, EventArgs e)
+            dte.StatusBar.Text = text;
+        }
+
         private static string GetStatusBarText(string packageName, string? installedVersion)
             => installedVersion == null ? packageName : $"{packageName} - {installedVersion}";
 
-        private bool TryUninstallPackage(
-            string packageName, Guid projectGuid, EnvDTE.DTE dte, EnvDTE.Project dteProject)
+        private async Task<bool> TryUninstallPackageAsync(
+            string packageName, Guid projectGuid, EnvDTE.DTE dte, EnvDTE.Project dteProject,
+            IProgressTracker progressTracker, CancellationToken cancellationToken)
         {
-            this.AssertIsForeground();
             Contract.ThrowIfFalse(IsEnabled);
 
-            // TODO: consider putting this under a TWD so it would actually be cancellable.
-            var cancellationToken = CancellationToken.None;
+            // explicitly switch to BG thread to do the installation as nuget installer APIs are free threaded.
+            await TaskScheduler.Default;
+
+            var description = string.Format(ServicesVSResources.Uninstalling_0, packageName);
+            progressTracker.Description = description;
+            await UpdateStatusBarAsync(dte, description, cancellationToken).ConfigureAwait(false);
 
             try
             {
-                return this.ThreadingContext.JoinableTaskFactory.Run(() =>
+                return await this.PerformNuGetProjectServiceWorkAsync(async (nugetService, cancellationToken) =>
                 {
-                    return this.PerformNuGetProjectServiceWorkAsync(async (nugetService, cancellationToken) =>
-                    {
-                        await this.ThreadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+                    var installedPackagesMap = await GetInstalledPackagesMapAsync(nugetService, projectGuid, cancellationToken).ConfigureAwait(true);
+                    if (!installedPackagesMap.TryGetValue(packageName, out var installedVersion))
+                        return false;
 
-                        var installedPackagesMap = await GetInstalledPackagesMapAsync(nugetService, projectGuid, cancellationToken).ConfigureAwait(true);
-                        if (!installedPackagesMap.TryGetValue(packageName, out var installedVersion))
-                            return false;
+                    cancellationToken.ThrowIfCancellationRequested();
 
-                        dte.StatusBar.Text = string.Format(ServicesVSResources.Uninstalling_0, packageName);
-                        _packageUninstaller.Value.UninstallPackage(dteProject, packageName, removeDependencies: true);
+                    // Once we start the installation, we can't cancel anymore.
+                    cancellationToken = default;
+                    _packageUninstaller.Value.UninstallPackage(dteProject, packageName, removeDependencies: true);
```

Comment:
❗ This requires a preceding call to switch to the main thread until https://github.com/NuGet/Home/issues/11021 is complete. Adding explicit switches to the main thread for operations that are not yet documented will allow this refactoring to proceed without being blocked, and the subsequent change once the blocking issue is addressed will be minimal.

Reply:
➡️ No change required here once NuGet/NuGet.Client#4209 merges.